File: src/0350-Intersection-of-Two-Arrays-II/0350.py
class Solution:
    def intersect(self, nums1, nums2):
        """
        :type nums1: List[int]
        :type nums2: List[int]
        :rtype: List[int]
        """
        record, result = {}, []
        for num in nums1:
            record[num] = record.get(num, 0) + 1
                
        for num in nums2:
            if num in record and record[num]:
                result.append(num)
                record[num] -= 1
        return result



if __name__ == "__main__":
    nums1 = [1, 2, 2, 1]
    nums2 = [2, 2]
    print(Solution().intersect(nums1, nums2))

File: src/0350-Intersection-of-Two-Arrays-II/0350.cpp
#include <unordered_map>
#include <iostream>
#include <vector>
using namespace std;

static int x = []() {std::ios::sync_with_stdio(false); cin.tie(0); return 0; }();
class Solution
{
public:
    vector<int> intersect(vector<int>& nums1, vector<int>& nums2)
    {
        unordered_map<int, int> record;
        vector<int> result;
        for (auto& num : nums1) ++record[num];
        for (auto& num : nums2)
        {
            if (record[num] > 0)
            {
                result.push_back(num);
                --record[num];
            }
        }
        return result;
    }
};
int main()
{
    vector<int> nums1 = { 1, 2, 2, 1 };
    vector<int> nums2 = { 2, 2 };
    vector<int> result = Solution().intersect(nums1, nums2);
    for (auto& i : result)
    {
        cout << i << " ";
    }
    cout << endl;
    return 0;
}